pkg/utils/fsutils/license_scanning_fs.go

```diff
@@ -0,0 +1,292 @@
+		if !ok {
+			return xerrors.Errorf("type assertion error, filepath: %s", path)
+		}
+		if readable, err := licenseutils.IsHumanReadable(file, math.MaxInt); err != nil || !readable {
+			return nil
+		}
+
+		content, err := io.ReadAll(file)
+		if err != nil {
+			return xerrors.Errorf("unable to read file content; %q: %w", path, err)
+		}
+
+		w.processChan <- licensing.ClassifierInput{
+			PkgID:               pkgID,
+			FilePath:            path,
+			Content:             content,
+			ConfidenceLevel:     w.ClassifierConfidenceLevel,
+			LicenseTextCacheDir: w.LicenseTextCacheDir,
+		}
+		return nil
+	}
+
+	// Walk through every file present in given directory except dependency dir and manifest file
+	if err := WalkDir(fsys, root, required, classifier); err != nil {
+		w.Logger.Error("walkDir utils has failed", log.String("root", root), log.String("error", err.Error()))
+	}
+
+	foundPackageDependencyDir := checkPackageDependencyDir(fsys, root, w.PackageDependencyDir)
+	if foundPackageDependencyDir {
+		return w.handlePackageDependencies(fsys, root, pkgID)
+	}
+
+	return true, nil
+}
+
+// For special packages (like scoped packages in npm), we need to recurse further for scanning
+func (w *RecursiveWalker) handleSpecialPath(fsys fs.FS, root string) (bool, error) {
+	dirEntries, err := fs.ReadDir(fsys, root)
+	if err != nil {
+		w.Logger.Error("failed to read dir contents", log.String("err", err.Error()))
+		return false, xerrors.Errorf("failed to read dir contents: %w", err)
+	}
+
+	for _, dirEntry := range dirEntries {
+		if dirEntry.IsDir() {
+			dependencyPath := path.Join(root, dirEntry.Name())
+			if ret, err := w.Walk(fsys, dependencyPath, ""); !ret || err != nil {
+				w.Logger.Error("Recursive walker has failed", log.String("path", dependencyPath))
+			}
+		}
+	}
+	return true, nil
+}
+
+// parses the package manifest file if present and valid. generates declared license
+func (w *RecursiveWalker) processPackageManifest(fsys fs.FS, root, parentPkgID string) (string, error) {
+	packageManifestPath := path.Join(root, w.PackageManifestFile)
+
+	if root != "." && w.PackageManifestFile != "" {
+		// check if package Manifest file exists, if yes, then parse then we parse it
+		if f, err := fs.Stat(fsys, packageManifestPath); err == nil && f.Size() != 0 {
+			pkg, err := w.Parser.ParseManifest(fsys, packageManifestPath)
+			if err != nil {
+				w.Logger.Error("unable to parse package manifest", log.String("path", packageManifestPath), log.String("error", err.Error()))
+				return "", xerrors.Errorf("unable to parse package manifest: %w", err)
+			}
+
+			w.Logger.Debug("Found declared licenses", log.String("pkgID", pkg.PackageID()),
+				log.Any("declaredLicenses", pkg.DeclaredLicenses()))
+
+			w.mutex.Lock()
+			w.licenses[pkg.PackageID()] = append(w.licenses[pkg.PackageID()], pkg.DeclaredLicenses()...)
+			w.mutex.Unlock()
+
+			return pkg.PackageID(), nil
+		}
+	}
+
+	var pkgID string
+	if parentPkgID == "" {
```

Comment:
this is confusing and doesn't seem right. Same variable return a string constant and a package id (name/version), which have a totally different meaning.

Reply:
I agree. changing it, please check